src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,240 @@
+from __future__ import annotations
+import dataclasses
+import datetime as dt
+import json
+import logging
+from collections.abc import Iterable, Sequence
+from typing import ClassVar, Protocol, TypeVar, Generic, Any, get_type_hints
+
+from databricks.labs.lsql.backends import SqlBackend
+
+from databricks.labs.ucx.framework.owners import Ownership
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.progress.install import Historical
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassWithIdAttributes(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+
+    __id_attributes__: ClassVar[tuple[str, ...]]
+    """The names of attributes (can be dataclass fields or ordinary properties) that make up the object identifier.
+
+    All attributes must be (non-optional) strings.
+    """
+
+
+Record = TypeVar("Record", bound=DataclassWithIdAttributes)
+
+
+class HistoricalEncoder(Generic[Record]):
+    """An encoder for dataclasses that will be stored in our history log.
+
+    Our history records are designed with several goals in mind:
+     - Records can be of different types, meaning we have to store heterogenous types using a homogenous schema.
+     - We partially shred the records to allow for easier SQL-based querying.
+     - Flexibility from the start, because schema changes in the future will be very difficult.
+     - Records will be shared (and queried) across workspaces with different versions of UCX installed.
```

Comment:
@nfx : Suggested that we need to enforce the same UCX version across workspaces (somehow)

Reply:
I need a bit more context on that… can you point me to this?

At the very least any enforcement needs to be relaxed enough to allow for upgrades.

Reply:
Let's move this conversation out of this PR https://github.com/databrickslabs/ucx/issues/3043